static_frame/core/index_hierarchy.py

```diff
@@ -140,202 +277,252 @@ def from_product(cls: tp.Type[IH],
+                    yield (label, *treerow)
+            else:
+                for row in subtree:
+                    yield (label, row)
 
     @classmethod
     def from_tree(cls: tp.Type[IH],
             tree: TreeNodeT,
             *,
             name: NameType = None,
-            index_constructors: tp.Optional[IndexConstructors] = None,
+            index_constructors: IndexConstructors = None,
             ) -> IH:
         '''
         Convert into a ``IndexHierarchy`` a dictionary defining keys to either iterables or nested dictionaries of the same.
 
         Returns:
             :obj:`static_frame.IndexHierarchy`
         '''
-        return cls(
-                cls._LEVEL_CONSTRUCTOR.from_tree(
-                        tree,
-                        index_constructors
-                        ),
+        return cls.from_labels(
+                labels=cls._from_tree(tree),
                 name=name,
+                index_constructors=index_constructors,
+                )
+
+    @classmethod
+    def _from_empty(cls: tp.Type[IH],
+            empty_labels: tp.Iterable[tp.Sequence[tp.Hashable]],
+            *,
+            name: NameType = None,
+            depth_reference: tp.Optional[int] = None,
+            ) -> IH:
+        '''
+        Construct an IndexHierarchy from an iterable of empty labels.
+        '''
+        if isinstance(empty_labels, np.ndarray) and empty_labels.ndim == 2:
+            # if this is a 2D array, we can get the depth
+            depth = empty_labels.shape[1]
+
+            if depth == 0: # an empty 2D array can have 0 depth
+                pass # do not set depth_reference, assume it is set
+            elif (depth_reference is None and depth > 1) or (
+                depth_reference is not None and depth_reference == depth
+            ):
+                depth_reference = depth
+            else:
+                raise ErrorInitIndex(
+                    f'depth_reference provided {depth_reference} does not match depth of supplied array {depth}'
+                )
+
+        if not isinstance(depth_reference, INT_TYPES):
+            raise ErrorInitIndex(
+                'depth_reference must be an integer when labels are empty.'
+            )
+
+        if depth_reference == 1:
+            raise ErrorInitIndex('Cannot create IndexHierarchy from only one level.')
+
+        return cls(
+                indices=[
+                    cls._INDEX_CONSTRUCTOR(EMPTY_TUPLE) for _ in range(depth_reference)
+                ],
+                indexers=[PositionsAllocator.get(0) for _ in range(depth_reference)],
+                name=name
+                )
+
+    @classmethod
+    def _from_array(cls: tp.Type[IH],
+            array: np.ndarray,
+            *,
+            name: NameType = None,
+            depth_reference: tp.Optional[DepthLevelSpecifier] = None,
+            index_constructors: IndexConstructors = None,
+            ) -> IH:
```

Comment:
Usage of this method is problematic, as there are few cases (I expect) when each depth is the same dtype; and callers might be tempted to use `.values` to call this type, which might in inadvertently coerce types. Are you sure we need it?